src/backend/distributed/operations/worker_shard_copy.c

```diff
@@ -0,0 +1,494 @@
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "libpq-fe.h"
+#include "postgres.h"
+#include "commands/copy.h"
+#include "nodes/makefuncs.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "utils/builtins.h"
+#include "distributed/remote_commands.h"
+#include "distributed/worker_shard_copy.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/worker_manager.h"
+#include "distributed/connection_management.h"
+#include "distributed/relation_utils.h"
+#include "distributed/version_compat.h"
+#include "distributed/local_executor.h"
+
+/*
+ * LocalCopyBuffer is used in copy callback to return the copied rows.
+ * The reason this is a global variable is that we cannot pass an additional
+ * argument to the copy callback.
+ */
+static StringInfo LocalCopyBuffer;
+
+typedef struct ShardCopyDestReceiver
+{
+	/* public DestReceiver interface */
+	DestReceiver pub;
+
+	/* Destination Relation Name */
+	List *destinationShardFullyQualifiedName;
+
+	/* descriptor of the tuples that are sent to the worker */
+	TupleDesc tupleDescriptor;
+
+	/* state on how to copy out data types */
+	CopyOutState copyOutState;
+	FmgrInfo *columnOutputFunctions;
+
+	/* number of tuples sent */
+	int64 tuplesSent;
+
+	/* destination node id */
+	uint32_t destinationNodeId;
+
+	/* local copy if destination shard in same node */
+	bool useLocalCopy;
+
+	/* EState for per-tuple memory allocation */
+	EState *executorState;
+
+	/*
+	 * Connection for destination shard (NULL if useLocalCopy is true)
+	 */
+	MultiConnection *connection;
+} ShardCopyDestReceiver;
+
+static bool ShardCopyDestReceiverReceive(TupleTableSlot *slot, DestReceiver *dest);
+static void ShardCopyDestReceiverStartup(DestReceiver *dest, int operation,
+										 TupleDesc inputTupleDescriptor);
+static void ShardCopyDestReceiverShutdown(DestReceiver *destReceiver);
+static void ShardCopyDestReceiverDestroy(DestReceiver *destReceiver);
+static bool CanUseLocalCopy(uint64 destinationNodeId);
+static StringInfo ConstructCopyStatement(List *destinationShardFullyQualifiedName, bool
+										 useBinaryFormat);
+static void WriteLocalTuple(TupleTableSlot *slot, ShardCopyDestReceiver *copyDest);
+static bool ShouldSendCopyNow(StringInfo buffer);
+static int ReadFromLocalBufferCallback(void *outBuf, int minRead, int maxRead);
+static void LocalCopyToShard(ShardCopyDestReceiver *copyDest, CopyOutState
+							 localCopyOutState);
+static void ConnectToRemoteAndStartCopy(ShardCopyDestReceiver *copyDest);
+
+static bool
+CanUseLocalCopy(uint64 destinationNodeId)
+{
+	/* If destination node is same as source, use local copy */
+	return GetLocalNodeId() == destinationNodeId;
+}
+
+
+/*
+ * ShouldSendCopyNow returns true if the given buffer size exceeds the
+ * local copy buffer size threshold.
+ */
+static bool
+ShouldSendCopyNow(StringInfo buffer)
```

Comment:
it still bugs me a bit that this function appears multiple times, but maybe the comparison shouldn't be in a separate function